Solution for decay products of Long-lived exotic particles in FastSim

sbein opened the pull request:
#### PR description:
Note: The proposed changes to the code have no effect on SM physics by construction, nor on existing prompt BSM models. The changes impact events where an exotic particle (pdgId>1000000) has a displaced decay with dxy>~3cm. 

Problem: 
BAU, some decay products of  long-lived (LL) particles are not simulated by FastSim.

Explanation
FastSim decays displaced particles (dxy(origin, PV)>xy(beam pipe)) "one by one" by creating an empty Pythia event, loading in the one particle, and using pythia8::next() to decay. However, some exotic particles cannot be decayed "on their own", for example, LL top squarks - in reality, these would form R-hadrons, but we don't always generate those. In such cases, FastSim just gives up and no decay products are produced.

Proposed solution:
Let’s make it default to take pre-set decay products of exotics. This will be the same as what's done for FullSim by default. 
Another option would be to compel anyone requesting such models to work with realistic particles, say, R-hadrons in place of LL squarks. But, this would reduce FastSim's flexibility. 

#### PR validation:

Validation has been performed by comparing equivalent FastSim and FullSim recipes. These validation studies were presented 
https://indico.cern.ch/event/949596/contributions/3993708/attachments/2093307/3517762/FastSimAugust2020Update.pdf

#### 
These changes should also be backported to previous releases 9_4_X, 10_2_X, and 10_6_X.

backport 9_4_X #31575
fix backport 9_4_X #31596
backport 10_2_X #31597
backport 10_6_X #31598

sbein committed:
overriding daughter and granddaughter exotic decay products generated by the FastSim Decayer with those in the original event record

sbein committed:
updated the check for exotic relatives to include grandmothers/granddaughters

sbein committed:
added a 3rd generation of exceptions for exotics

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/17985

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/17985/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/17985/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
post-code checks commit

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/17987

- This PR adds an extra 20KB to repository

cmsbuild commented:
A new Pull Request was created by  @sbein (Sam Bein) for master.

It involves the following packages:

FastSimulation/SimplifiedGeometryPropagator


@ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please review it and eventually sign? Thanks.
@matt-komm this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

sbein commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b43c8a3043973a124af23dee8b3d89a0d88d3f7a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/8990/summary.html
CMSSW: CMSSW_11_2_X_2020-08-28-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

davidlange6 commented on FastSimulation/SimplifiedGeometryPropagator/src/Decayer.cc:
```diff
@@ -33,6 +33,15 @@ void fastsim::Decayer::decay(const Particle& particle,
   int pid = particle.pdgId();
+  // snip decay products of exotic particles or their children. These decay products are preserved from the event record.
+  // limitation: if exotic incurs heavy energy loss during propagation, the saved decay products could be too hard.
+  if (abs(pid) > 1000000) {
```
please use std::abs here and elsewhere

sbein replied:
done

davidlange6 commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -212,9 +212,63 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+            const HepMC::GenParticle& genRelative3rdGen = **relatives3rdGenIterator_;
+            if (abs(genRelative3rdGen.pdg_id()) > 1000000) {
+              exoticRelativeId =
+                  0;  //the current particle, if non-stable, will have its decay products remade by fastsim
```
code checks made this rather unreadable. you might rearrange the comment..

sbein replied:
addressed this by making comment its own line

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/8990/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2609638
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

sbein commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

sbein committed:
Update ParticleManager.cc

fixing comment according to David's comment

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update Decayer.cc

using std::abs instead of abs

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update ParticleManager.cc

switching to std::abs

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18001

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18001/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18001/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
Update ParticleManager.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18002

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18002/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18002/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
trying to apply code formatting again after I removed a problematic comment by hand and then it was rejected

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18003

- This PR adds an extra 24KB to repository

sbein commented:
+1

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1
Tested at: 09afc3ee562227e3e4e9dedb12efdde43ba266e3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/8995/summary.html
CMSSW: CMSSW_11_2_X_2020-08-28-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/8995/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2609667
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2609633
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -212,9 +212,62 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+    int exoticRelativeId = 0;
+    if (productionVertex->position().perp2() * lengthUnitConversionFactor2_ > beamPipeRadius2_)  //
+    {
+      std::vector<HepMC::GenParticle*>::const_iterator relativesIterator_ =
+          productionVertex->particles_in_const_begin();
+      std::vector<HepMC::GenParticle*>::const_iterator relativesIteratorEnd_ =
+          productionVertex->particles_in_const_end();
+      bool hasExoticAssociation = false;
+      for (; relativesIterator_ != relativesIteratorEnd_; ++relativesIterator_) {
+        const HepMC::GenParticle& genRelative = **relativesIterator_;
```
I think the modern loop is more readable (and shorter)
```
      bool hasExoticAssociation = false;
      for (const auto &relativesIterator_ : productionVertex->particles_in(particle)) {
        const HepMC::GenParticle& genRelative = *relativesIterator_;
```

If you use this version, you need to add `#include "HepMC/GenRanges.h"` in `ParticleManager.h`

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -212,9 +212,62 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+        if (!relVertex)
+          continue;
+
+        std::vector<HepMC::GenParticle*>::const_iterator relatives2ndGenIterator_ =
+            relVertex->particles_in_const_begin();
+        std::vector<HepMC::GenParticle*>::const_iterator relatives2ndGenIteratorEnd_ =
+            relVertex->particles_in_const_end();
+        for (; relatives2ndGenIterator_ != relatives2ndGenIteratorEnd_; ++relatives2ndGenIterator_) {
+          const HepMC::GenParticle& genRelative2ndGen = **relatives2ndGenIterator_;
```
same thing here

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -212,9 +212,62 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+          const HepMC::GenVertex* relVertex2 = genRelative2ndGen.production_vertex();
+          if (!relVertex2)
+            continue;
+          std::vector<HepMC::GenParticle*>::const_iterator relatives3rdGenIterator_ =
+              relVertex2->particles_in_const_begin();
+          std::vector<HepMC::GenParticle*>::const_iterator relatives3rdGenIteratorEnd_ =
+              relVertex2->particles_in_const_end();
+          for (; relatives3rdGenIterator_ != relatives3rdGenIteratorEnd_; ++relatives3rdGenIterator_) {
+            const HepMC::GenParticle& genRelative3rdGen = **relatives3rdGenIterator_;
```
...and here

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -212,9 +212,62 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+            break;
+          }
+          const HepMC::GenVertex* relVertex2 = genRelative2ndGen.production_vertex();
+          if (!relVertex2)
+            continue;
+          std::vector<HepMC::GenParticle*>::const_iterator relatives3rdGenIterator_ =
+              relVertex2->particles_in_const_begin();
+          std::vector<HepMC::GenParticle*>::const_iterator relatives3rdGenIteratorEnd_ =
+              relVertex2->particles_in_const_end();
+          for (; relatives3rdGenIterator_ != relatives3rdGenIteratorEnd_; ++relatives3rdGenIterator_) {
+            const HepMC::GenParticle& genRelative3rdGen = **relatives3rdGenIterator_;
+            if (std::abs(genRelative3rdGen.pdg_id()) > 1000000) {
+              exoticRelativeId = 0;
+              hasExoticAssociation = true;
+              break;
+            }
```
Instead of repeating nested code it would be much better if you define a function or something like that.
This would make the code simpler, safer, and ready to be extended to get the N-th relative.

sbein replied:
I can attempt this. It is a bit strange to me because they are fundamentally nested loops, but there may be a clever way to implement these loops with functions. If there is not a concrete suggestion, I will attempt something and get back.

silviodonato commented:
-1
(just to remind to myself there are some pending comments)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

sbein committed:
built in a recursive function into the exotic relatives checker loop

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18078

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

civanch commented:
@sbein , may be the main condition to identify exotic particle in this update should be changed, because ions will be called "exotic" which is not the case. FOr example, PDG code for alpha is 1000020040.

civanch commented:
@sbein , we have a more complex condition:

bool Generator::isExotic(int pdgcode) const {
  int pdgid = std::abs(pdgcode);
  return ((pdgid >= 1000000 && pdgid < 4000000 && pdgid != 3000022) ||  // SUSY, R-hadron, and technicolor particles
          pdgid == 17 ||                                                // 4th generation lepton
          pdgid == 34 ||                                                // W-prime
          pdgid == 37);                                                 // charged Higgs
}

sbein committed:
added check of vertex object in iterative particle relative checker, extended the number of generations to 100, and adopted Vladimir's definition of an exotic particle

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18110

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

silviodonato commented:
@sbein please define a very simple function to check the condition
`(relpid >= 1000000 && relpid < 4000000 && relpid != 3000022) || relpid == 17 || relpid == 34 || relpid == 37`
and use this function in the code instead of repeating 5 times the same condition.
This will make the code more easy more maintainable and readble.

Please remove also the comment ` // previously: particle must be produced within the beampipe `. We can use the github history to check the previous code.

sbein committed:
skipping propagation for quarks and gluons, since they can somehow sneak into the acceptance

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18374

- This PR adds an extra 32KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18374/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18374/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
ran formatting code check

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18375

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

sbein commented:
The nested loops were exchanged for a function, and this allowed the number of generations searched for exotics to be extended - it is now set at 100. Further tests with long-lived particles revealed very rare seg faults related to FastSim attempting to propagate (displaced) quarks themselves and do material interactions, via a divide by particle mass (divide by zero) in EnergyLoss.cc. This is obviously not desired, so an additional line has been added in ParticleManager.cc that explicitly skips the propagation of quarks and gluons.

civanch commented on FastSimulation/SimplifiedGeometryPropagator/src/Decayer.cc:
```diff
@@ -33,6 +33,19 @@ void fastsim::Decayer::decay(const Particle& particle,
+  // limitation: if exotic incurs heavy energy loss during propagation, the saved decay products could be too hard.
+  int abspid = std::abs(pid);
+  if ((abspid >= 1000000 && abspid < 4000000 && abspid != 3000022) || abspid == 17 || abspid == 34 || abspid == 37) {
+    std::cout << "not decaying due to pdgid=" << pid << std::endl;
```
@sbein , std::cout should not be used, you need edm::LogVerbatim

civanch commented:
@sbein , in FullSim we also filter out quarks and gluons - in g4SimHits there is a vector of unwanted pdg IDs, which may be configured. I also follow Silvio suggestion to make a filter method (may be even inlined), so filter code will be implemented in one place only.

sbein committed:
defined inline function for determination of exotic association

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18403

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

sbein commented:
+1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/Decayer.cc:
```diff
@@ -33,6 +34,16 @@ void fastsim::Decayer::decay(const Particle& particle,
+  // snip decay products of exotic particles or their children. These decay products are preserved from the event record.
+  // limitation: if exotic incurs heavy energy loss during propagation, the saved decay products could be too hard.
+
+  if (isExotic(pid)) {
+    return;
+  }
+  if (isExotic(particle.getMotherPdgId())) {
+    return;
+  }
```
```suggestion
if (isExotic(pid) || isExotic(particle.getMotherPdgId())){
    return;
}
```

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/interface/ParticleManager.h:
```diff
@@ -140,4 +141,9 @@ namespace fastsim {
   };
 }  // namespace fastsim
 
+inline bool isExotic(int pdgid_) {
```
Please add a comment explaining the function and the numbers (eg. `3000022`).

sbein replied:
added comments here from Vladimir

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -211,11 +211,19 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+    if (productionVertex->position().perp2() * lengthUnitConversionFactor2_ > beamPipeRadius2_)  //
+    {
+      exoticRelativesChecker(productionVertex, exoticRelativeId, 0);
+      bool hasExoticAssociation = isExotic(exoticRelativeId);
+      if (!hasExoticAssociation) {
```
```suggestion
if (!isExotic(exoticRelativeId))
```

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -267,3 +278,32 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+void fastsim::ParticleManager::exoticRelativesChecker(const HepMC::GenVertex* originVertex,
+                                                      int& exoticRelativeId_,
+                                                      int ngendepth = 0) {
+  int relpid = std::abs(exoticRelativeId_);
+  bool relIsExotic = isExotic(relpid);
+
+  if (ngendepth > 99 || exoticRelativeId_ == -1 || relIsExotic)
```
```suggestion
   if (ngendepth > 99 || exoticRelativeId_ == -1 || isExotic(std::abs(exoticRelativeId_))) 
```

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -267,3 +278,32 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
+  std::vector<HepMC::GenParticle*>::const_iterator relativesIteratorEnd_ = originVertex->particles_in_const_end();
+  for (; relativesIterator_ != relativesIteratorEnd_; ++relativesIterator_) {
+    const HepMC::GenParticle& genRelative = **relativesIterator_;
+    relpid = std::abs(genRelative.pdg_id());
+    relIsExotic = isExotic(relpid);
+    if (relIsExotic) {
```
```suggestion
    if (isExotic(std::abs(genRelative.pdg_id()))) {
```

silviodonato commented on FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:
```diff
@@ -267,3 +278,32 @@ std::unique_ptr<fastsim::Particle> fastsim::ParticleManager::nextGenParticle() {
 
   return std::unique_ptr<Particle>();
 }
+
```
Please add a comment to describe the function

silviodonato reviewed (commented):
I've added some minor comments

cmsbuild commented:
+1
Tested at: b3e2e32c9893f794111284d02d4e662fae525105
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/9399/summary.html
CMSSW: CMSSW_11_2_X_2020-09-17-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

sbein committed:
Update FastSimulation/SimplifiedGeometryPropagator/src/Decayer.cc

Co-authored-by: Silvio Donato <silvio.donato@cern.ch>

sbein committed:
Update FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc

Co-authored-by: Silvio Donato <silvio.donato@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc

Co-authored-by: Silvio Donato <silvio.donato@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc

Co-authored-by: Silvio Donato <silvio.donato@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update ParticleManager.h

taking labeling as suggested by Vladimir

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein commented:
please test

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:228:9: error: use of undeclared identifier 'endVertex' [clang-diagnostic-error]
    if (endVertex && endVertex->position().perp2() * lengthUnitConversionFactor2_ < beamPipeRadius2_) {
        ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:228:22: error: use of undeclared identifier 'endVertex' [clang-diagnostic-error]
    if (endVertex && endVertex->position().perp2() * lengthUnitConversionFactor2_ < beamPipeRadius2_) {
                     ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:229:7: error: 'continue' statement not in loop statement [clang-diagnostic-error]
      continue;
      ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:234:22: error: use of undeclared identifier 'particle' [clang-diagnostic-error]
        new Particle(particle.pdg_id(),
                     ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:235:46: error: use of undeclared identifier 'productionVertex' [clang-diagnostic-error]
                     math::XYZTLorentzVector(productionVertex->position().x() * lengthUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:236:46: error: use of undeclared identifier 'productionVertex' [clang-diagnostic-error]
                                             productionVertex->position().y() * lengthUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:237:46: error: use of undeclared identifier 'productionVertex' [clang-diagnostic-error]
                                             productionVertex->position().z() * lengthUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:238:46: error: use of undeclared identifier 'productionVertex' [clang-diagnostic-error]
                                             productionVertex->position().t() * timeUnitConversionFactor_),
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:239:46: error: use of undeclared identifier 'particle' [clang-diagnostic-error]
                     math::XYZTLorentzVector(particle.momentum().x() * momentumUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:240:46: error: use of undeclared identifier 'particle' [clang-diagnostic-error]
                                             particle.momentum().y() * momentumUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:241:46: error: use of undeclared identifier 'particle' [clang-diagnostic-error]
                                             particle.momentum().z() * momentumUnitConversionFactor_,
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:242:46: error: use of undeclared identifier 'particle' [clang-diagnostic-error]
                                             particle.momentum().e() * momentumUnitConversionFactor_)));
                                             ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:244:18: error: use of undeclared identifier 'exoticRelativeId' [clang-diagnostic-error]
    if (isExotic(exoticRelativeId)) {
                 ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:245:35: error: use of undeclared identifier 'exoticRelativeId' [clang-diagnostic-error]
      newParticle->setMotherPdgId(exoticRelativeId);
                                  ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:249:9: error: unknown type name 'endVertex'; did you mean 'SimVertex'? [clang-diagnostic-error]
    if (endVertex) {
        ^~~~~~~~~
        SimVertex
--
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:249:18: error: expected unqualified-id [clang-diagnostic-error]
    if (endVertex) {
                 ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:251:12: error: use of undeclared identifier 'endVertex' [clang-diagnostic-error]
          (endVertex->position().t() - productionVertex->position().t()) * timeUnitConversionFactor_;
           ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:251:40: error: use of undeclared identifier 'productionVertex' [clang-diagnostic-error]
          (endVertex->position().t() - productionVertex->position().t()) * timeUnitConversionFactor_;
                                       ^
FastSimulation/SimplifiedGeometryPropagator/src/ParticleManager.cc:278:3: error: expected unqualified-id [clang-diagnostic-error]
  return std::unique_ptr<Particle>();
  ^
Suppressed 659 warnings (657 in non-user code, 2 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/9399/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2540471
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2540448
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

sbein committed:
Update ParticleManager.cc

fixing missing bracket after accepting changes

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18439

- This PR adds an extra 80KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18439/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18439/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
Update ParticleManager.cc

tweaking code formatting things by hand

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update Decayer.cc

tweaking more brackets by hand

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein commented:
please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18440

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18440/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18440/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
Update ParticleManager.h

And more indentation code format checks that i have to do by hand

sbein commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18441

- This PR adds an extra 88KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18441/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18441/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

sbein committed:
Update ParticleManager.cc

More subtle formatting issues I'm forced to track down by hand

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein committed:
Update Decayer.cc

More tiny changes

cmsbuild commented:
The code-checks are being triggered in jenkins.

sbein commented:
please test

sbein committed:
Update Decayer.cc

Hopefully the last thing I have to change by hand, given the original release is now deprecated

sbein commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31285/18444

- This PR adds an extra 32KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #31285 was updated. @ssekmen, @lveldere, @civanch, @mdhildreth, @cmsbuild, @sbein can you please check and sign again.

cmsbuild commented:
+1
Tested at: b17c4053fc4b71690e5e09fb15f866ff3af8b4ea
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/9409/summary.html
CMSSW: CMSSW_11_2_X_2020-09-17-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

sbein commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-259aa4/9409/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2540471
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2540442
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

silviodonato commented:
Summary of the backports
backport 9_4_X #31575
fix backport 9_4_X #31596
backport 10_2_X #31597
backport 10_2_X #31598
